Есть таблица истории просмотров фильмов в онлайн-кинотеатре.
Два поля: Пользователь, Фильм
Множество строк: тысячи пользователей и фильмов.

Нужно сделать SQL запрос или скрипт на Python,
который бы реализовал логику алгоритма рекомендаций фильмов к просмотру, которые пользователи еще не смотрели.
Логику рекомендаций предлагайте на свое усмотрение.
На выходе нужна таблица: Пользователь, Рекомендованный фильм, Рейтинг рекомендации


Поскольку в задаче нет дополнительной информации о фильмах и пользователях, которую можно было бы анализировать(описание фильмов, жанры, год выпуска, возраст зрителя и т.д.) будем реализовывать алгоритм, основанный на коллаборативной фильтрации (**Collaborative filtering**). 

Суть состоит в том, чтобы вычислить степень попарной "похожести" фильмов друг на друга. **Чем больше пользователей, посмотревших один из фильмов, посмотрело оба фильма** тем больше фильмы похожи. Таким образом для N фильмов получаем матрицу NxN, которую используем для формирования рекомендаций для пользователя по его истории просмотров.     

In [357]:
import numpy as np

In [358]:
import pandas as pd

In [359]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [360]:
# lets use MovieLens latest small dataset for testing: https://grouplens.org/datasets/movielens/ 
data = pd.read_csv('/content/drive/MyDrive/ratings.csv')[['movieId', 'userId', 'rating']]

In [361]:
data

movieId  userId  rating
0             1       1     4.0
1             3       1     4.0
2             6       1     4.0
3            47       1     5.0
4            50       1     5.0
...         ...     ...     ...
100831   166534     610     4.0
100832   168248     610     5.0
100833   168250     610     5.0
100834   168252     610     5.0
100835   170875     610     3.0

[100836 rows x 3 columns]

In [362]:
# For perfomance we take first 1000 movies
mIds = data['movieId'].unique()[:1000]
data = data[data['movieId'].isin(mIds)]

In [363]:
data

movieId  userId  rating
0             1       1     4.0
1             3       1     4.0
2             6       1     4.0
3            47       1     5.0
4            50       1     5.0
...         ...     ...     ...
100672   109374     610     4.5
100673   109487     610     3.5
100732   119145     610     4.5
100737   122882     610     5.0
100771   136020     610     3.5

[44399 rows x 3 columns]

In [364]:
#split data on test & train
from sklearn.model_selection import train_test_split
train_data, test_data = train_test_split(data, test_size= 0.3)

for each movie find their userIds and their count
then calculate intersections in loop

In [365]:
#for each movie find users who watched it and count them
def get_reviewers(data):
    keys, values = data[['movieId', 'userId']].sort_values('movieId').values.T
    ukeys, index = np.unique(keys, True)
    arrays = np.split(values, index[1:])
    df = pd.DataFrame({'movieId':ukeys, 'usersSet':[set(a) for a in arrays], 'userCount':[len(a) for a in arrays]})
    return df, ukeys

In [366]:
movieViewers, k = get_reviewers(train_data)

In [367]:
# form similarity matrix to make predictions
similarityMatrix = pd.DataFrame(index=list(range(len(k))), columns=list(range(len(k))), dtype=float)

In [368]:
similarityMatrix

0    1    2    3    4    5    6    ...  984  985  986  987  988  989  990
0    NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN
1    NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN
2    NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN
3    NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN
4    NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN
..   ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...
986  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN
987  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN
988  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN
989  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN
990  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN  NaN  NaN  NaN  NaN

[991 rows x 991 columns]

In [369]:
#to compute similarity jaccard index is used https://en.wikipedia.org/wiki/Jaccard_index
def jaccard_index(x, movie_viewers):
    B = movie_viewers.values[x.name]
    B_count = B[2]
    B_set = B[1]
    if (B_count == 0):
      return np.zeros(shape=(len(x.index),))
    res = np.zeros(shape=(len(x.index),))
    for i in x.index:
        if(i >= x.name):
          #compute only top diagonal part for performance 
          return res
        A = movie_viewers.values[i]
        A_count = A[2]
        A_set = A[1]
        if (A_count == 0):
            res[i] = 0.0
        else:
            intersection = len(A_set & B_set)
            res[i] = intersection / (A_count + B_count - intersection)

    return res

In [370]:
#!pip install swifter

In [371]:
#import swifter

In [372]:
#compute similarity matrix
res2 = similarityMatrix.apply(jaccard_index, args=(movieViewers,))

In [373]:
#fill bottom diagonal
res2 = res2.values.T + res2.values
np.fill_diagonal(res2, 1.0)

In [374]:
#similarityMatrix = pd.DataFrame(res2.values, index=k, columns=k)
similarityMatrix = pd.DataFrame(res2, index=k, columns=k)

In [375]:
similarityMatrix

1         2         3       ...    136020    137595    140110
1       1.000000  0.176166  0.115607  ...  0.024242  0.000000  0.012346
2       0.176166  1.000000  0.142857  ...  0.073171  0.000000  0.037500
3       0.115607  0.142857  1.000000  ...  0.018868  0.000000  0.000000
4       0.006289  0.025974  0.022727  ...  0.000000  0.000000  0.000000
5       0.103448  0.144330  0.203125  ...  0.039216  0.000000  0.021277
...          ...       ...       ...  ...       ...       ...       ...
119145  0.041667  0.032967  0.034483  ...  0.090909  0.047619  0.068966
122882  0.074286  0.091837  0.013889  ...  0.166667  0.000000  0.128205
136020  0.024242  0.073171  0.018868  ...  1.000000  0.066667  0.190476
137595  0.000000  0.000000  0.000000  ...  0.066667  1.000000  0.100000
140110  0.012346  0.037500  0.000000  ...  0.190476  0.100000  1.000000

[991 rows x 991 columns]

In [376]:
#add movies not included in train data
similarityMatrix = pd.DataFrame(similarityMatrix, index=mIds, columns=mIds).fillna(0.0)

In [377]:
similarityMatrix.isnull().values.any()

False

In [378]:
similarityMatrix

1         3         6     ...      2100      2421      2485
1     1.000000  0.115607  0.143590  ...  0.109195  0.073620  0.060976
3     0.115607  1.000000  0.161290  ...  0.200000  0.111111  0.035088
6     0.143590  0.161290  1.000000  ...  0.069307  0.071429  0.034884
47    0.197581  0.123457  0.145946  ...  0.103030  0.058065  0.031646
50    0.180000  0.090909  0.206897  ...  0.090909  0.065789  0.038710
...        ...       ...       ...  ...       ...       ...       ...
1876  0.094118  0.126984  0.052083  ...  0.145161  0.104167  0.061224
2072  0.043210  0.125000  0.024390  ...  0.125000  0.161290  0.093750
2100  0.109195  0.200000  0.069307  ...  1.000000  0.200000  0.092593
2421  0.073620  0.111111  0.071429  ...  0.200000  1.000000  0.078947
2485  0.060976  0.035088  0.034884  ...  0.092593  0.078947  1.000000

[1000 rows x 1000 columns]

In [379]:
def predict(views: np.ndarray, similarity: np.ndarray):
    pred = (views.dot(similarity) / similarity.sum(axis=1))
    return np.nan_to_num(pred, nan=0.0)

EVALUATION


In [380]:
test_users = test_data['userId'].unique()
total_watched = 0
right_preds = 0
result_df = pd.DataFrame(columns=['userId', 'movieId', 'rating'])
for user_id in test_users:
    #get movies user watched in test data
    test_user_watched = test_data[test_data['userId'] == user_id]['movieId'].values
    #get movies user already watched
    user_movie_indx = train_data[train_data['userId'] == user_id]['movieId'].values
    user_movie_rating = train_data[train_data['userId'] == user_id]['rating'].values
    #create vector with movie ids and 1
    user_watched = pd.DataFrame(np.ones(shape=(len(user_movie_indx),)), index = user_movie_indx)
    #fill not watched movies with zeros
    user_watched = pd.DataFrame(user_watched, index=mIds).fillna(0.0)
    #get predictions for user
    pred = predict(user_watched.values.reshape((len(user_watched.values),)), similarityMatrix.values)
    train_user_watched = train_data[train_data['userId'] == user_id]['movieId'].values
    #get user recommendations
    pred = pd.DataFrame(pred, index=mIds, columns=['rating'])
    #filter already watched films
    pred = pred[~pred.index.isin(train_user_watched)]

    tmp = pd.DataFrame(pred, columns=['userId', 'movieId', 'rating'])
    tmp['movieId'] = tmp.index
    tmp['userId'] = user_id
    result_df = result_df.append(tmp, ignore_index=True)

    #get top N recommendations for user with N views
    pred = pred.sort_values('rating',ascending = False).head(len(test_user_watched))
    #print(pred[pred.index.isin(test_user_watched)].shape[0],len(test_user_watched))
    total_watched += len(test_user_watched)
    right_preds +=  pred[pred.index.isin(test_user_watched)].shape[0]

print(right_preds / total_watched)
print(result_df)

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:2: RuntimeWarning: invalid value encountered in true_divide
  


0.36906906906906906
       userId movieId    rating
0         599     151  0.496229
1         599     316  0.592860
2         599     362  0.481788
3         599     423  0.413867
4         599     527  0.576236
...       ...     ...       ...
575939    499    1876  0.006340
575940    499    2072  0.010178
575941    499    2100  0.007067
575942    499    2421  0.007260
575943    499    2485  0.006068

[575944 rows x 3 columns]


Таким образом модель предсказала 37% просмотренных фильмов из тестового набора.

Недостатком подхода является необходимость досчитывать матрицу при добавлении новых фильмов и невозможность рекоммендовать фильмы пользователям без истории просмотров. 

Аналогичную модель можно построить на поиске похожих пользователей по истории просмотров. 

